subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java

```diff
@@ -65,7 +66,9 @@ public R execute(UnitOfWork work, AfterPreviousExecutionContext context) {
         ImmutableCollection<String> warnings = problems.get(Severity.WARNING);
         ImmutableCollection<String> errors = problems.get(Severity.ERROR);
 
-        warnings.forEach(warningReporter::reportValidationWarning);
+        if (!warnings.isEmpty()) {
```

Comment:
I would put the `isEmpty()` check on the `warningReporter` side, so the validation step can also call `reportValidationWarnings()`.

Reply:
Can you explain why/how that would be better?

Reply:
It seems to me that the `warningReporter` should decide whether or not there are any warnings. I don't feel super strong about it.